### PR TITLE
Structured error handling

### LINKED ISSUE #81: Sanitize error handling
- Fix error response
- Introduce error codes and error message table
- Place error translator in dropwizard
- Handle errors and return properly

### PR DESCRIPTION
**Pull request by r0goyal**
Fixes #81

### CODE REVIEW

**Comment by r0goyal** [comment]
@santanusinha Please review this

File: foxtrot-common/src/main/java/com/flipkart/foxtrot/common/query/FilterOperator.java [hunk 1]
```diff
@@ -22,24 +22,24 @@
  */
 public interface FilterOperator {
     //All
-    public static final String equals = "equals";
-    public static final String not_equals = "not_equals";
-    public static final String any = "any";
-    public static final String in = "in";
-    public static final String exists = "exists";
-    public static final String missing = "missing";
+    String equals = "equals";
```
Comment (santanusinha) [inline foxtrot-common/src/main/java/com/flipkart/foxtrot/common/query/FilterOperator.java]: Eh?!! Why are these not static final? These are fixed strings.. You're basically screwing the compiler...

File: foxtrot-core/src/main/java/com/flipkart/foxtrot/core/exception/FoxtrotException.java [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+
+    public ErrorCode getCode() {
+        return code;
+    }
+
+    public void setCode(ErrorCode code) {
+        this.code = code;
+    }
+
+    public static TableInitializationException createTableInitializationException(Table table, String message) {
```
Comment (santanusinha) [inline foxtrot-core/src/main/java/com/flipkart/foxtrot/core/exception/FoxtrotException.java]: Won't it be better if you create another class called Exceptions and move the create functions there?

File: foxtrot-core/src/main/java/com/flipkart/foxtrot/core/exception/DocumentMissingException.java [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+/**
+ * Created by rishabh.goyal on 18/12/15.
+ */
+public class DocumentMissingException extends FoxtrotException {
+
+    private String table;
+    private List<String> ids;
+
+    public DocumentMissingException(String table, List<String> ids) {
```
Comment (santanusinha) [inline foxtrot-core/src/main/java/com/flipkart/foxtrot/core/exception/DocumentMissingException.java]: You have create* functions in the base class, even if you move them out to another class (see comment on base class), these should be package local as only the create* functions will access this.

File: foxtrot-core/src/main/java/com/flipkart/foxtrot/core/querystore/actions/GroupAction.java [hunk 4]
```diff
@@ -112,30 +123,29 @@ public ActionResponse execute(GroupRequest parameter) throws QueryStoreException
...
     }
 
     private Map<String, Object> getMap(List<String> fields, Aggregations aggregations) {
         final String field = fields.get(0);
         final List<String> remainingFields = (fields.size() > 1) ? fields.subList(1, fields.size())
-                : new ArrayList<String>();
+                : new ArrayList<>();
         Terms terms = aggregations.get(Utils.sanitizeFieldForAggregation(field));
-        Map<String, Object> levelCount = new HashMap<String, Object>();
+        Map<String, Object> levelCount = new HashMap<>();
```
Comment (santanusinha) [inline foxtrot-core/src/main/java/com/flipkart/foxtrot/core/querystore/actions/GroupAction.java]: Seriously? This and not Maps.createHashMap()?

File: foxtrot-core/src/main/java/com/flipkart/foxtrot/core/querystore/impl/ElasticsearchUtils.java [hunk 5]
```diff
@@ -127,101 +127,101 @@ public static PutIndexTemplateRequest getClusterTemplateMapping(IndicesAdminClie
...
-                                                    .startObject()
-                                                        .field("format", "doc_values")
-                                                    .endObject()
-                                            .endObject()
-                                        .endObject()
-                                .endObject()
-                            .endArray()
-                    .endObject()
-            .endObject();
+                .startObject()
```
Comment (santanusinha) [inline foxtrot-core/src/main/java/com/flipkart/foxtrot/core/querystore/impl/ElasticsearchUtils.java]: indent... The readability is zero on this.

**Comment by santanusinha** [comment]
In general 2 comments:
1) We should move to guava for creates
2) Create a global exception handler at jersey level to return proper exception message

File: foxtrot-common/src/main/java/com/flipkart/foxtrot/common/query/FilterOperator.java (hunk 1 above)
Comment (r0goyal) [inline foxtrot-common/src/main/java/com/flipkart/foxtrot/common/query/FilterOperator.java]: Interface variables are by default static and final.

File: foxtrot-core/src/main/java/com/flipkart/foxtrot/core/querystore/actions/GroupAction.java (hunk 4 above)
Comment (r0goyal) [inline foxtrot-core/src/main/java/com/flipkart/foxtrot/core/querystore/actions/GroupAction.java]: This particular guava method seems redundant with Java 7 feature allowing to create map without specifying parameters. Any particular advantage of using Guava method ?

**Comment by r0goyal** [comment]
@santanusinha 2nd point is already done -> Check following class - FoxtrotExceptionMapper

File: foxtrot-common/src/main/java/com/flipkart/foxtrot/common/query/FilterOperator.java (hunk 1 above)
Comment (santanusinha) [inline foxtrot-common/src/main/java/com/flipkart/foxtrot/common/query/FilterOperator.java]: Ah..  Missed the interface...

File: foxtrot-core/src/main/java/com/flipkart/foxtrot/core/querystore/actions/GroupAction.java (hunk 4 above)
Comment (santanusinha) [inline foxtrot-core/src/main/java/com/flipkart/foxtrot/core/querystore/actions/GroupAction.java]: Smaller code...

**Comment by santanusinha** [comment]
My bad... Missed it.

**Comment by santanusinha** [comment]
Fix the other comments about exception creation and we can merge.

**Comment by r0goyal** [comment]
@santanusinha Changes done. As for mapping indentation, every time I format code, it becomes distorted. Will try to figure out some way of properly indenting it in next refactoring PR

File: foxtrot-core/src/main/java/com/flipkart/foxtrot/core/exception/ExceptionUtils.java [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+import com.flipkart.foxtrot.common.ActionRequest;
+import com.flipkart.foxtrot.common.Table;
+
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * Created by rishabh.goyal on 03/01/16.
+ */
+public class ExceptionUtils {
```
Comment (santanusinha) [inline foxtrot-core/src/main/java/com/flipkart/foxtrot/core/exception/ExceptionUtils.java]: One minor gripe.. can you rename this to FoxtrotExceptions

**Pull request merged**